README
Crypt-OpenSSL-ECDSA version 0.03s
================================

This module provides an interface to the ECDSA (Elliptic Curve Digital
Signature Algorithm) functions in OpenSSL.


INSTALLATION

To compile and install this module type the following:

   perl Makefile.PL
   make
   make test
   make install

By default, this will compile against and link with OpenSSL in
/usr/include/openssl and /usr/lib/libcrypto.so

You can change the location of the OpenSSL headers and library to use with the
OPENSSL_PREFIX (which defaults to /usr)
environment variable, for example:

OPENSSL_PREFIX=/home/mikem/playpen/openssl-1.0.2/ perl Makefile.PL


DEPENDENCIES

This module requires these other modules and libraries:

 Crypt::OpenSSL::Bignum
 Crypt::OpenSSL::EC
 OpenSSL (0.9.8a or later) libcrypto.so library and headers with EC and ECDSA enabled

COPYRIGHT AND LICENCE

Copyright (C) 2014 by Mike McCauley

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.18.1 or,
at your option, any later version of Perl 5 you may have available.




Changes
Revision history for Perl extension Crypt::OpenSSL::ECDSA.

0.01  Tue Oct 21 07:02:43 2014
      - original version; created by h2xs 1.23 with options
		-n Crypt::OpenSSL::ECDSA /usr/include/openssl/ecdsa.h -L /usr/lib -lcrypt

0.02  Oct 21 14:41 
      - Added missing typemap

0.03  Oct 31 14:36
      - ECDSA_METHOD functions are only available with OpenSSL 1.0.2 and later
      - ECDSA_SIG is now a blessed object
      - Added ECDSA_SIG get_r set_r get_s get_s functions and tests
      - Documentation still to be done

0.04 Jan 2 2015
      - In ECDSA.pm, use version restriction eased to 5.6.2
      - Added documentation for get_s and get_r, courtesy Dick Franks.
      - Fixed leaks in set_r and set_s

0.05 Jan 3 2015
      - Added Crypt::OpenSSL::EC as a prerequisite.


lib/Crypt/OpenSSL/ECDSA.pm
package Crypt::OpenSSL::ECDSA;

use 5.008005;
use strict;
use warnings;
use Carp;

require Exporter;
use AutoLoader;

our @ISA = qw(Exporter);

# Items to export into callers namespace by default. Note: do not export
# names by default without a very good reason. Use EXPORT_OK instead.
# Do not simply export all your public functions/methods/constants.

# This allows declaration	use Crypt::OpenSSL::ECDSA ':all';
# If you do not need this, moving things directly into @EXPORT or @EXPORT_OK
# will save memory.
our %EXPORT_TAGS = ( 'all' => [ qw(
	ECDSA_F_ECDSA_CHECK
	ECDSA_F_ECDSA_DATA_NEW_METHOD
	ECDSA_F_ECDSA_DO_SIGN
	ECDSA_F_ECDSA_DO_VERIFY
	ECDSA_F_ECDSA_SIGN_SETUP
	ECDSA_R_BAD_SIGNATURE
	ECDSA_R_DATA_TOO_LARGE_FOR_KEY_SIZE
	ECDSA_R_ERR_EC_LIB
	ECDSA_R_MISSING_PARAMETERS
	ECDSA_R_NEED_NEW_SETUP_VALUES
	ECDSA_R_NON_FIPS_METHOD
	ECDSA_R_RANDOM_NUMBER_GENERATION_FAILED
	ECDSA_R_SIGNATURE_MALLOC_FAILED
) ] );

our @EXPORT_OK = ( @{ $EXPORT_TAGS{'all'} } );

our @EXPORT = qw(
	ECDSA_F_ECDSA_CHECK
	ECDSA_F_ECDSA_DATA_NEW_METHOD
	ECDSA_F_ECDSA_DO_SIGN
	ECDSA_F_ECDSA_DO_VERIFY
	ECDSA_F_ECDSA_SIGN_SETUP
	ECDSA_R_BAD_SIGNATURE
	ECDSA_R_DATA_TOO_LARGE_FOR_KEY_SIZE
	ECDSA_R_ERR_EC_LIB
	ECDSA_R_MISSING_PARAMETERS
	ECDSA_R_NEED_NEW_SETUP_VALUES
	ECDSA_R_NON_FIPS_METHOD
	ECDSA_R_RANDOM_NUMBER_GENERATION_FAILED
	ECDSA_R_SIGNATURE_MALLOC_FAILED
);

our $VERSION = '0.05';

sub AUTOLOAD {
    # This AUTOLOAD is used to 'autoload' constants from the constant()
    # XS function.

    my $constname;
    our $AUTOLOAD;
    ($constname = $AUTOLOAD) =~ s/.*:://;
    croak "&Crypt::OpenSSL::ECDSA::constant not defined" if $constname eq 'constant';
    my ($error, $val) = constant($constname);
    if ($error) { croak $error; }
    {
	no strict 'refs';
	# Fixed between 5.005_53 and 5.005_61
#XXX	if ($] >= 5.00561) {
#XXX	    *$AUTOLOAD = sub () { $val };
#XXX	}
#XXX	else {
	    *$AUTOLOAD = sub { $val };
#XXX	}
    }
    goto &$AUTOLOAD;
}


eval {
	local $SIG{'__DIE__'} = 'DEFAULT';

	eval {
		require XSLoader;
		XSLoader::load( 'Crypt::OpenSSL::ECDSA', $VERSION );
		1;
	} or do {
		use vars qw(@ISA);
		require DynaLoader;
		push @ISA, 'DynaLoader';
		bootstrap('Crypt::OpenSSL::ECDSA', $VERSION);
	};
};


# Preloaded methods go here.

# Autoload methods go after =cut, and are processed by the autosplit program.

1;
__END__

=head1 NAME

Crypt::OpenSSL::ECDSA - Perl extension for OpenSSL ECDSA (Elliptic Curve Digital Signature Algorithm)

=head1 SYNOPSIS

    use Crypt::OpenSSL::ECDSA;
    $sig = Crypt::OpenSSL::ECDSA::ECDSA_do_sign( $digest, $eckey );
    $r = $sig->get_r;
    $s = $sig->get_s;

    $sig = Crypt::OpenSSL::ECDSA::ECDSA_SIG->new();
    $sig->set_r($r);
    $sig->set_s($s);
    Crypt::OpenSSL::ECDSA::ECDSA_do_verify( $digest, $sig, $eckey );

=head1 DESCRIPTION

This module provides an interface to the ECDSA (Elliptic Curve Digital
Signature Algorithm) functions in OpenSSL

Tested against OpenSSL 1.0.2


=head2 Export

None by default.


=head2 Exportable constants

    ECDSA_F_ECDSA_CHECK
    ECDSA_F_ECDSA_DATA_NEW_METHOD
    ECDSA_F_ECDSA_DO_SIGN
    ECDSA_F_ECDSA_DO_VERIFY
    ECDSA_F_ECDSA_SIGN_SETUP
    ECDSA_R_BAD_SIGNATURE
    ECDSA_R_DATA_TOO_LARGE_FOR_KEY_SIZE
    ECDSA_R_ERR_EC_LIB
    ECDSA_R_MISSING_PARAMETERS
    ECDSA_R_NEED_NEW_SETUP_VALUES
    ECDSA_R_NON_FIPS_METHOD
    ECDSA_R_RANDOM_NUMBER_GENERATION_FAILED
    ECDSA_R_SIGNATURE_MALLOC_FAILED


=head1 CLASS METHODS

=item new

    $sig = Crypt::OpenSSL::ECDSA::ECDSA_SIG->new();

Create a new ECDSA Signature Object.

The internal "r" and "s" components are undefined.


=head1 OBJECT METHODS

=item get_r;

    $r = $sig->get_r;

Gets signature "r" component.

The method returns a string representing a binary integer in network
(big-endian) byte order.


=item get_s;

    $s = $sig->get_s;

Gets signature "s" component.

The method returns a string representing a binary integer in network
(big-endian) byte order.


=item set_r;

    $sig->set_r($r);

Sets signature "r" component.

The argument is a string representing a binary integer in network
(big-endian) byte order.


=item set_s;

    $sig->set_r($s);

Sets signature "s" component.

The argument is a string representing a binary integer in network
(big-endian) byte order.

=head1 SEE ALSO

L<OpenSSL documentation|http://www.openssl.org/>,
L<Crypt::OpenSSL::EC>,

=head1 AUTHOR

Mike McCauley, E<lt>mikem@airspayce.comE<gt>


=head1 COPYRIGHT AND LICENSE

Copyright (C) 2014 by Mike McCauley

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.18.1 or,
at your option, any later version of Perl 5 you may have available.

=cut


Makefile.PL
use 5.018001;
use ExtUtils::MakeMaker;

# You might need to set this if your openssl headers and libs anre in an unusual place:
my $openssl_prefix = $ENV{OPENSSL_PREFIX} || '/usr';

# See lib/ExtUtils/MakeMaker.pm for details of how to influence
# the contents of the Makefile that is written.
WriteMakefile(
    NAME              => 'Crypt::OpenSSL::ECDSA',
    VERSION_FROM      => 'lib/Crypt/OpenSSL/ECDSA.pm', # finds $VERSION
    PREREQ_PM         => {'Crypt::OpenSSL::Bignum' => 0.04,
			  'Crypt::OpenSSL::EC' => 0.05,
                         },
    ($] >= 5.005 ?     ## Add these new keywords supported since 5.005
      (ABSTRACT_FROM  => 'lib/Crypt/OpenSSL/ECDSA.pm', # retrieve abstract from module
       AUTHOR         => 'Mike McCauley <mikem@airspayce.com>') : ()),
    LIBS              => ["-L$openssl_prefix/lib -lcrypto"], 
    DEFINE            => '-DPERL5 -DOPENSSL_NO_KRB5', 
    INC               => "-I. -I$openssl_prefix/include",
	# Un-comment this if you add C files to link with later:
    # OBJECT            => '$(O_FILES)', # link all the C files too
);
if  (eval {require ExtUtils::Constant; 1}) {
  # If you edit these definitions to change the constants used by this module,
  # you will need to use the generated const-c.inc and const-xs.inc
  # files to replace their "fallback" counterparts before distributing your
  # changes.
  my @names = (qw(ECDSA_F_ECDSA_CHECK ECDSA_F_ECDSA_DATA_NEW_METHOD
		 ECDSA_F_ECDSA_DO_SIGN ECDSA_F_ECDSA_DO_VERIFY
		 ECDSA_F_ECDSA_SIGN_SETUP ECDSA_R_BAD_SIGNATURE
		 ECDSA_R_DATA_TOO_LARGE_FOR_KEY_SIZE ECDSA_R_ERR_EC_LIB
		 ECDSA_R_MISSING_PARAMETERS ECDSA_R_NEED_NEW_SETUP_VALUES
		 ECDSA_R_NON_FIPS_METHOD
		 ECDSA_R_RANDOM_NUMBER_GENERATION_FAILED
		 ECDSA_R_SIGNATURE_MALLOC_FAILED));
  ExtUtils::Constant::WriteConstants(
                                     NAME         => 'Crypt::OpenSSL::ECDSA',
                                     NAMES        => \@names,
                                     DEFAULT_TYPE => 'IV',
                                     C_FILE       => 'const-c.inc',
                                     XS_FILE      => 'const-xs.inc',
                                  );

}
else {
  use File::Copy;
  use File::Spec;
  foreach my $file ('const-c.inc', 'const-xs.inc') {
    my $fallback = File::Spec->catfile('fallback', $file);
    copy ($fallback, $file) or die "Can't copy $fallback to $file: $!";
  }
}
